main/main.c

```diff
@@ -1,67 +1,240 @@
+        // not sure if this should restart, but also not sure what else to do here?
+        esp_restart();
+    }
+
+#if CONFIG_LORA_TRANSMITTER
+    initialize_sensors();
+    xTaskCreate(&task_tx, "TX", 1024 * 3, NULL, 5, NULL);
+#endif
+
+#if CONFIG_LORA_RECEIVER && CONFIG_ENABLE_MQTT
+    initialize_mqtt();
+#endif
+
+#if CONFIG_LORA_RECEIVER
+    xTaskCreate(&task_rx, "RX", 1024 * 3, NULL, 5, NULL);
+#endif
+}
+
+#if CONFIG_LORA_TRANSMITTER
+void task_tx(void *pvParameters)
+{
+    ESP_LOGI(pcTaskGetName(NULL), "Start TX");
+    vTaskDelay(500 / portTICK_PERIOD_MS);
+
+    packet_t packet = {0};
+
+    while (1)
+    {
+        double temp_raw, press_raw, hum_raw;
+        bme280_read_temperature(&temp_raw);
+        bme280_read_pressure(&press_raw);
+        bme280_read_humidity(&hum_raw);
+
+        // Fill the packet with the meta data
+        packet.version = (CONFIG_PACKET_VERSION << 4) | 0; // Reserved 4 bits set to 0
+        packet.id = (CONFIG_CLASS_ID << 4) | CONFIG_DEVICE_ID;
+        packet.msgID = 1;                   // Example message ID
+        packet.msgCount = 1;                // Example message count (optional, set as needed)
+        packet.dataType = CONFIG_DATA_TYPE; // Example data type
+
+        if (packet.dataType == BME280)
+        {
+            // Convert raw sensor readings
+            packet.data[0] = (int8_t)(temp_raw * 2);             // Scale temperature for higher precision and fit into int8_t
+            packet.data[1] = (int8_t)((press_raw / 100) - 1000); // Convert Pa to hPa, subtract 1000
+            packet.data[2] = (uint8_t)hum_raw;                   // Fit humidity into uint8_t
+
+            // Log the packet data before sending
+            ESP_LOGI(pcTaskGetName(NULL), "Temperature: %.2f C (scaled to %d)", temp_raw, packet.data[0]);
+            ESP_LOGI(pcTaskGetName(NULL), "Pressure: %.2f hPa (stored as %d)", press_raw / 100, packet.data[1]);
+            ESP_LOGI(pcTaskGetName(NULL), "Humidity: %.2f %% (stored as %u)", hum_raw, packet.data[2]);
+        }
+        else if (packet.dataType == SMS)
+        {
+            // pack a String "BuzzVerse" into data array
+            char data[] = "BuzzVerse";
+            for (int i = 0; i < sizeof(data); i++)
+            {
+                packet.data[i] = data[i];
+            }
+
+            // Log the packet data before sending
+            ESP_LOGI(pcTaskGetName(NULL), "Data: %s", packet.data);
+        } // Add more data types here
+
+        lora_status_t send_status = lora_send(&packet);
+
+        if (LORA_OK != send_status)
+        {
+            ESP_LOGE(TAG, "Packet send failed");
+        }
+        else
+        {
+            ESP_LOGI(TAG, "Packet sent successfully");
+        }
+
+        // low_power_mode_set_sleep_time(15 * 60); // 15 minutes
+        // low_power_mode_enter_deep_sleep();
+    }
+}
+
+#endif
+
+#if CONFIG_LORA_RECEIVER
+void task_rx(void *pvParameters)

```

Comment:
Split for smaller functions maybe?

Reply:
Same story as TX but, pass queue as parameter and then if data is received then put it into the queue.